sw/device/tests/sysrst_ctrl_inputs_test.c

```diff
@@ -0,0 +1,154 @@
+OTTF_DEFINE_TEST_CONFIG(.enable_uart_flow_control = true);
+
+static dif_sysrst_ctrl_t sysrst_ctrl;
+const uint32_t kNumPhases = 10;
+
+// On DV, we must use variables in flash but on a real device,
+// we must use variables in RAM.
+static const volatile uint8_t kTestPhaseDV = 0;
+static const volatile uint8_t kTestExpectedDV = 0;
+// In DV, the sequence can ensure that the pins are set even before the test
+// runs. On a real device, this is not the case and if the initial value of
+// kTestPhase is 0, the very first OTTF_WAIT_FOR could succeed before the host
+// can set the pins. To avoid this, and only on real devices, set the initial
+// value to an invalid value so that we have to wait for the host.
+static volatile uint8_t kTestPhaseReal = 0xff;
+static volatile uint8_t kTestExpectedReal = 0;
+
+enum {
+  kOutputNumPads = 0x8,
+  kOutputNunMioPads = 0x6,
+};
+
+static const dif_pinmux_index_t kPeripheralInputs[] = {
+    kTopEarlgreyPinmuxPeripheralInSysrstCtrlAonKey0In,
+    kTopEarlgreyPinmuxPeripheralInSysrstCtrlAonKey1In,
+    kTopEarlgreyPinmuxPeripheralInSysrstCtrlAonKey2In,
+    kTopEarlgreyPinmuxPeripheralInSysrstCtrlAonPwrbIn,
+    kTopEarlgreyPinmuxPeripheralInSysrstCtrlAonAcPresent,
+    kTopEarlgreyPinmuxPeripheralInSysrstCtrlAonLidOpen,
+};
+
+static const dif_pinmux_index_t kInputPadsDV[] = {
+    kTopEarlgreyPinmuxInselIob3, kTopEarlgreyPinmuxInselIob6,
+    kTopEarlgreyPinmuxInselIob8, kTopEarlgreyPinmuxInselIor13,
+    kTopEarlgreyPinmuxInselIoc7, kTopEarlgreyPinmuxInselIoc9,
+};
+
+static const dif_pinmux_index_t kInputPadsReal[] = {
+    kTopEarlgreyPinmuxInselIor2, kTopEarlgreyPinmuxInselIor3,
+    kTopEarlgreyPinmuxInselIor4, kTopEarlgreyPinmuxInselIor5,
+    kTopEarlgreyPinmuxInselIor6, kTopEarlgreyPinmuxInselIor7,
+};
+
+static const dif_sysrst_ctrl_pin_t kSysrstCtrlInputs[] = {
+    kDifSysrstCtrlPinKey0In,           kDifSysrstCtrlPinKey1In,
+    kDifSysrstCtrlPinKey2In,           kDifSysrstCtrlPinPowerButtonIn,
+    kDifSysrstCtrlPinAcPowerPresentIn, kDifSysrstCtrlPinLidOpenIn,
+    kDifSysrstCtrlPinEcResetInOut,     kDifSysrstCtrlPinFlashWriteProtectInOut,
+};
+
+static uint8_t read_input_pins(void) {
+  bool input_value;
+  uint8_t inputs = 0;
+  for (int i = 0; i < kOutputNumPads; ++i) {
+    CHECK_DIF_OK(dif_sysrst_ctrl_input_pin_read(
+        &sysrst_ctrl, kSysrstCtrlInputs[i], &input_value));
+    inputs |= input_value << i;
+  }
+  return inputs;
+}
+
+void setup_dio_pins(dif_pinmux_t *pinmux, dif_sysrst_ctrl_t *sysrst_ctrl) {
+  // Make sure the the DIO pins EC reset and flash WP are configured correctly.
+  // The documentation says that they should be configured as open drain but on
+  // FPGA targets, only virtual open drain is supported.
+  dif_pinmux_pad_attr_t out_attr;
+  dif_pinmux_pad_attr_t in_attr = {
+      .slew_rate = 0,
+      .drive_strength = 0,
+      // FIXME FPGA requires virtual open drain, should be open drain on real
+      // chip?
```

Comment:
That's correct. sysrst_ctrl is a block where either side can affect the outputs, and the DIOs are specifically open-drain type, to allow both to pull low.

For the FPGA with hyperdebug, the "correct" configuration would be to have hyperdebug's corresponding GPIOs set to `PinMode::OpenDrain` and `PullMode::PullUp` (if that's possible -- I haven't checked yet!).

Reply:
Is that just for the EC reset and flash WP pins? I am current using
`transport.gpio_pin(pin)?.set_mode(PinMode::PushPull)?;` on all pins.

Reply:
For the MIOs on _this test_, it's technically not necessary to do open drain mode, since the outputs are never activated, but I believe in practice, the I/O standard for most of these pins is typically open-drain (even for the pads that are only capable of virtual open-drain).

For the DIOs (EC reset and flash WP), it's needed to avoid the chance of having two chips driving against each other.